watertap/tools/parallel/parallel_manager.py

```diff
@@ -0,0 +1,155 @@
+from abc import abstractmethod, ABC
+import os
+
+import numpy
+
+from pyomo.common.dependencies import attempt_import
+
+
+class ParallelManager(ABC):
+    """
+    Create and return an instance of a ParallelManager, based on the libraries available in the
+    runtime environment.
+
+    Allows an optional python class to be passed in as parallel_manager_class. If so, this class
+    is instantiated and returned rather than checking the local environment.
+    """
+
+    @classmethod
+    def create_parallel_manager(cls, parallel_manager_class=None):
+        if parallel_manager_class is not None:
+            return parallel_manager_class()
+
+        mpi4py, mpi4py_available = attempt_import("mpi4py")
+        if mpi4py_available and ParallelManager.is_running_under_mpi():
+            return MPIParallelManager(mpi4py)
+
+        return SingleProcessParallelManager()
+
+    """
+    Returns whether the process was run as part of an MPI group.
+    """
+
+    @classmethod
+    def is_running_under_mpi(cls):
+        # see https://www.open-mpi.org/faq/?category=running#mpi-environmental-variables.
+        # this is one of the environment variables that MPI sets.
+        return os.getenv("OMPI_COMM_WORLD_SIZE") is not None
+
+    """
+    Return whether the current process should be considered as the root of the parallel group it
+    is a part of.
+    """
+
+    @abstractmethod
+    def is_root_process(self):
+        raise NotImplementedError
+
+    """
+    Synchronize a piece of data with all peer processes. The data parameter is either:
+    - (if root) the source of truth that will be broadcast to all processes
+    - (if not root) an empty buffer that will hold the data sent from root once sync_data returns
+    """
+
+    @abstractmethod
+    def sync_data(self, data):
+        raise NotImplementedError
+
+    """
+    Scatter a function out to a set of child processes, to be run
+    for a list of parameters.
+    - all_parameter_combos is a list, where each item represents the
+    parameters for a single run.
+    - fn is a function to be run, and should accept a single parameter. 
+    It will be called for each item in the all_parameter_combos list.
+    """
+
+    def scatter(self, all_parameter_combos, fn):
+        raise NotImplementedError
+
+    """
+    Gather the results of the computation that was kicked off via a previous
+    scatter.
+    - returns: an instance of GatherResults
+    """
+
+    def gather(self):
+        raise NotImplementedError
+
+
+"""
+Class representing the results of one process's run of an optimization routine.
+"""
+
+
+class LocalResults:
+    def __init__(self, parameters, results):
+        self.parameters = parameters
+        self.results = results
```

Comment:
I think since data classes use type annotations and we don't have those in watertap (yet?) it might throw people off. I can revisit if these classes get more complex though.

Reply:
That's fine, this is a soft suggestion.

To you other point, I would happily accept PRs with type annotations. Obviously not required, though perhaps the larger team should have that discussion.